### ISSUE TITLE
Lamachine update keeps failing

### ISSUE THREAD

**Issue by willstout** [proycon/LaMachine#94]
I'm updating Lamachine and I keep getting the message in red saying "The LaMachine update failed" with 464 as ok, 219 as changed, 0 as unreachable, and always 2 as failed. Can anyone tell me what's going on?

TASK [lamachine-git : labirinto - Checkout latest stable version v0.2.3] ********************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": false,
    "cmd": "git checkout v0.2.3",
    "delta": "0:00:00.178522",
    "end": "2018-06-13 16:37:54.441978",
    "rc": 1,
    "start": "2018-06-13 16:37:54.263456"
}

STDERR:

error: Your local changes to the following files would be overwritten by checkout:
        config/index.js
        config/prod.env.js
        package.json
Please commit your changes or stash them before you switch branches.
Aborting

MSG:

non-zero return code

TASK [lamachine-git : labirinto - STASHING AWAY LOCAL CHANGES!!!] ***************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": true,
    "cmd": "echo \"There appear to be local changes, LaMachine will stash them away\">&2\n echo \"this means they are not deleted and recoverable in principle,\">&2\n echo \"but they are NOT used in this compilation. You should \">&2\n echo \"git add them properly if you want to use any custom changes\">&2\n echo \"... allowing for 15 second grace period (CTRL-C in case of panic)...\"\n sleep 15\n git stash -u",
    "delta": "0:00:15.196583",
    "end": "2018-06-13 16:38:09.891121",
    "rc": 1,
    "start": "2018-06-13 16:37:54.694538"
}

STDOUT:

... allowing for 15 second grace period (CTRL-C in case of panic)...

STDERR:

There appear to be local changes, LaMachine will stash them away
this means they are not deleted and recoverable in principle,
but they are NOT used in this compilation. You should
git add them properly if you want to use any custom changes

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <(null)>) not allowed
Cannot save the current index state

MSG:

non-zero return code

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
localhost                  : ok=464  changed=219  unreachable=0    failed=2

======================================================================================
The LaMachine update failed! You have several options:
 - Retry a forced update (lamachine-update force=1), this forces recompilation even if software seems up to date
   and may be necessary in certain circumstances.
 - Retry the update, possibly tweaking configuration and installation options (lamachine-update --edit)
 - File a bug report on https://github.com/proycon/LaMachine/issues/
   The log file has been written to /usr/local/src/LaMachine/lamachine-latest-20180613_161753.log (include it with any bug report)

**Comment by proycon** [comment]
This looks like issue #88 , the fix I implemented didn't work properly it seems. I'll get on it asap!

**Comment by proycon** [comment]
I implemented another fix, could you perhaps check if your problem is resolved now with a normal ``lamachine-update`` (no force) ?

**Comment by willstout** [comment]
I’ll check tomorrow when I’m back at my office computer which has all my NLP stuff, I’ll be sure to get back to you then! Thanks for being so quick!

**Comment by willstout** [comment]
What kind of environment are you using to get your LaMachine to work? Also are you guys using Vagrant, Docker, etc...?

**Comment by proycon** [comment]
Just running ``lamachine-update`` was fine indeed; although it still seems to be stuck on the same thing as you initially reported. That's by fault I see, I applied the fix in one place but forgot it in another.

(As to your second question, the bug can pop up in any flavour (vagrant, docker, local virtual env) upon update.)

Could you give it another try with the latest fix?  (just run ``lamachine-update`` again, no force). It's a bit hard for me to reproduce exactly.

**Comment by willstout** [comment]
Still no luck sadly.

STDOUT:

v0.2.3

TASK [lamachine-git : labirinto - Checkout latest stable version v0.2.3] *************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": false,
    "cmd": "git checkout v0.2.3",
    "delta": "0:00:00.154589",
    "end": "2018-06-19 15:38:05.177024",
    "rc": 1,
    "start": "2018-06-19 15:38:05.022435"
}

STDERR:

error: Your local changes to the following files would be overwritten by checkout:
        config/index.js
        config/prod.env.js
        package.json
Please commit your changes or stash them before you switch branches.
Aborting


MSG:

non-zero return code

TASK [lamachine-git : labirinto - STASHING AWAY LOCAL CHANGES!!!] ********************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": true,
    "cmd": "echo \"There appear to be local changes, LaMachine will stash them away\">&2\n echo \"this means they are not deleted and recoverable in principle,\">&2\n echo \"but they are NOT used in this compilation. You should \">&2\n echo \"git add them properly if you want to use any custom changes\">&2\n echo \"... allowing for 15 second grace period (CTRL-C in case of panic)...\"\n sleep 15\n git stash -u",
    "delta": "0:00:15.222863",
    "end": "2018-06-19 15:38:20.637564",
    "rc": 1,
    "start": "2018-06-19 15:38:05.414701"
}

STDOUT:

... allowing for 15 second grace period (CTRL-C in case of panic)...


STDERR:

There appear to be local changes, LaMachine will stash them away
this means they are not deleted and recoverable in principle,
but they are NOT used in this compilation. You should
git add them properly if you want to use any custom changes

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <(null)>) not allowed
Cannot save the current index state


MSG:

non-zero return code

PLAY RECAP ***************************************************************************************************************************************************
localhost                  : ok=444  changed=173  unreachable=0    failed=2

======================================================================================
The LaMachine update failed! You have several options:
 - Retry a forced update (lamachine-update force=1), this forces recompilation even if software seems up to date
   and may be necessary in certain circumstances.
 - Retry the update, possibly tweaking configuration and installation options (lamachine-update --edit)
 - File a bug report on https://github.com/proycon/LaMachine/issues/
   The log file has been written to /usr/local/src/LaMachine/lamachine-latest-20180619_152840.log (include it with any bug report)

**Comment by proycon** [comment]
Strange, it looks as if the LaMachine scripts are still on an older version without my fix (which shouldn't be possible). Could you send me your entire log file? (``/usr/local/src/LaMachine/lamachine-latest-20180619_152840.log``)

**Comment by proycon** [comment]
Ah, I think I'm onto something. Are you using the docker version by any chance?

**Comment by proycon** [comment]
Assuming the docker update bug caused your issues, that should be fixed now. I just released a new LaMachine v2.2.7. However, you will need to obtain the image again from Docker Hub, as ``lamachine-update`` won't work in this case unfortunately.

**Comment by willstout** [comment]
Yeah I am using Docker

**Comment by willstout** [comment]
The update was successful! Do you still want the error log?

**Comment by willstout** [comment]
TL;DR: It seems as though it doesn't save the changes after it updating and then restarting. I'm not sure if this is something on my end or what. 

For more context I was attempting "ocr.nf" in lamachine and it told me that nextflow was missing some files, so I downloaded nextflow from the website and tried to put them in the right directory (as was nextflows instruction in command-prompt). After an hour of searching for the nextflow bin folder, which I still can't find, I decided to restart my computer out of frustration. Upon reboot, I typed "lamachine-update --edit" in command-prompt and found that none of my previous add-ons were kept (PICCL, Nextflow, etc...)

**Comment by proycon** [comment]
> TL;DR: It seems as though it doesn't save the changes after it updating and then restarting. I'm not sure if this is something on my end or what.

Yes, that's a docker thing, you'll get a new container every time, you need to explicitly commit any changes, effectively obtaining a new image. It's not a VM, you might be better of with one of the other flavours?

> For more context I was attempting "ocr.nf" in lamachine and it told me that nextflow was missing some files, so I downloaded nextflow from the website and tried to put them in the right directory (as was nextflows instruction in command-prompt). After an hour of searching for the nextflow bin folder, which I still can't find, I decided to restart my computer out of frustration. Upon reboot, I typed "lamachine-update --edit" in command-prompt and found that none of my previous add-ons were kept (PICCL, Nextflow, etc...)

Installation of PICCL, Nextflow is taken care of by LaMachine (after you explicitly opt-in for PICCL in the update). So if something is missing that could be a LaMachine bug. The other issue is a docker thing, which I can imagine may throw off people in the beginning, as mentioned.

**Comment by willstout** [comment]
As a follow up, you were obviously right about needing to commit changes with Docker, so that's working now. But the previous issue of trying to use "ocr.nf", then telling me I am missing a certain file, and then updating to try to get that file still persists. 

C:\Users\willstout>docker run -p 8080:80 -t -i proycon/lamachine:latest
==============================================================================
           ,              LaMachine v2.2.8 - Unified NLP Software distribution
          ~)    '                (https://proycon.github.io/LaMachine)
           (----)
            /| |\         Centre for Language and Speech Technology
           / / /|                    Radboud University Nijmegen
==============================================================================
 Build Name: latest
 Version: stable      Build time: 2018-06-21 16:04:29 UTC
 Maintainer: proycon@latest

Welcome to LaMachine!
  - run lamachine-list to see a verbose list of all installed software
     add -s for a short sorted view
  - run lamachine-update to update and test your LaMachine installation (provided you have sufficient permission)
     add the --edit flag if you want to install extra software.
  - run lamachine-start-webserver to (re)start the webserver (provided you have sufficient permission)

lamachine@a87ebb789431:/$ ocr.nf
ERROR ~ .nextflow/history.lock (No such file or directory)

 -- Check '.nextflow.log' file for details
lamachine@a87ebb789431:/$ nextflow self-update
mktemp: failed to create file via template '/nxf-tmp.XXXXXX': Permission denied
Downloading nextflow dependencies. It may require a few seconds, please wait .. curl: (23) Failed writing body (0 != 14840)
ERROR: Cannot download nextflow required file -- make sure you can connect to the internet

Alternatively you can try to download this file:
    https://www.nextflow.io/releases/latest/nextflow?a=update

and save it as:
    /usr/local/bin/nextflow

lamachine@a87ebb789431:/$ exit
logout

C:\Users\willstout>ping google.com

Pinging google.com [216.58.219.14] with 32 bytes of data:
Reply from 216.58.219.14: bytes=32 time=13ms TTL=49
Reply from 216.58.219.14: bytes=32 time=13ms TTL=49
Reply from 216.58.219.14: bytes=32 time=13ms TTL=49
Reply from 216.58.219.14: bytes=32 time=13ms TTL=49

Ping statistics for 216.58.219.14:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 13ms, Maximum = 13ms, Average = 13ms

**Comment by willstout** [comment]
Awesome, everything is working! You've been such a great help!
